### PR TITLE
Add permissions page

### PR DESCRIPTION
**Pull request by glacambre**
This commit adds a permission page written in inferno. This requires
several changes in the build process that were copied from #932 so
hopefully there won't be any merge conflicts.

The permission page allows granting Tridactyl more permissions. All of
Tridactyl's required permissions that could be made optional were made
optional. On top of that, all of Firefox's optional permissions were
added to the manifest.

There are multiple things left to do do:
- [ ] Decide if we really want to make every permission optional
- [ ] Implement a mechanism to prevent breakage when using an excmd
      which requires permissions that Tridactyl doesn't have
- [x] Add an easy way to navigate to the permissions page (e.g.
     `:permissions`, a link on the newtab page or the doc)
- [ ] Find a way to reuse the text we have in permissions.tsx in order
      to generate part of doc/amo.md and reduce duplication

Related issues: https://github.com/tridactyl/tridactyl/issues/788
https://github.com/tridactyl/tridactyl/issues/708

This will partially unblock https://github.com/tridactyl/tridactyl/pull/623 as the `find` permission is actually optional (discovered that while looking at Firefox's source code).

### CODE REVIEW

File: src/controller.ts [hunk 1]
```diff
@@ -110,9 +110,11 @@ export async function acceptExCmd(exstr: string): Promise<any> {
 }
 
 import { activeTabId } from "./lib/webext"
-browser.webNavigation.onBeforeNavigate.addListener(async function(details) {
-    if (details.frameId === 0 && details.tabId === (await activeTabId())) {
-        state.mode = "normal"
-    }
-})
+if (browser.webNavigation) {
```
Comment (bovine3dom) [inline src/controller.ts]: Could we have a proxy for browser that handled this stuff gracefully? e.g. `triBrowser` that prevents us from having to cover everything in ifs / care about whether a permission exists.

File: src/permissions.tsx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    (<Permission name="webRequest" reason="This is only needed if you set you use `:set csp clobber`."/>),
+    (<Permission name="webRequestBlocking" reason="Also needed in order to use `:set csp clobber`."/>),
+    // Non-optional permissions
+    (<Permission name="activeTab" reason="Required to inject CSS into tabs and access tab information."/>),
+    (<Permission name="browsingData" reason="Required in order to implement the `:sanitise` command."/>),
+    (<Permission name="contextualIdentities" reason="Necessary to interact with containers."/>),
+    (<Permission name="contextMenus"/>),
+    (<Permission name="sessions" reason="Used for restoring tabs/windows after they were closed."/>),
+    (<Permission name="storage" reason="Necessary to store your config in Tridactyl."/>),
+    (<Permission name="nativeMessaging" reason="Required in order to interact with your OS (start vim, read tridactylrc...)."/>),
```
Comment (bovine3dom) [inline src/permissions.tsx]: also opening about:* pages which is probably very unintuitive

**Comment by bovine3dom** [comment]
I don't quite understand the point of this. As I understand it, we can only ask for all of the optional permissions at once, or none of them.

E.g, clipboardWrite is a very different level of trust to history; and we have to have the native messenger permission which is the scariest of them all (although I guess it's pretty inert unless you install our binary).

I think this is adding an extra hoop for the 99.9% of Tridactyl users that are happy to give us all these permissions so that a tiny minority can have a nearly useless extension.

Or is your main aim to unblock #623 et al?

File: src/controller.ts (hunk 1 above)
Comment (glacambre) [reply src/controller.ts]: I thought about this but the proxy would need to be async (but we could work around that by prefetching and caching permissions) and another problem would be stopping code from executing.
If we throw, Tridactyl will crash just like when there's no proxy. If instead we return a recursive proxy we'll in the best cases crash just like when there's no proxy and in the worst cases silently fail.

**Comment by glacambre** [comment]
> As I understand it, we can only ask for all of the optional permissions at once, or none of them.

No, we can ask for permissions separately.

>Or is your main aim to unblock #623 et al?

That and I also wrote a firefox patch to turn every permission into an optional permission. I'll plan on submitting it soon.

**Comment by cmcaine** [comment]
Cool!

**Comment by bovine3dom** [comment]
Huh, I had that in my mind as a Fact. I guess I misread

> Only a single request is made for all requested permissions: thus either all permissions are granted or none of them are.

In which case this is much more sensible than I thought.

We should probably ask for all of the permissions on first launch, and have a link to the permissions page in all the usual places (new tab, help page, tutorial). Asking for them all might be a bit scary but I think it's better than the alternative of having Tridactyl be a bit useless (e.g, no history completions) for most users unless they do some digging.

**Comment by maturanomx** [comment]
On https://github.com/tridactyl/tridactyl/blob/master/src/static/clippy/settings.md#settings you could read:

> there will eventually be an options page where you can change and view these settings. For now, you can use viewconfig [option]. 

Maybe that text should be changed along with this PR? 🤔 

(I just discover this extension yesterday ... I'm loving it ❤️ )

**Comment by bovine3dom** [comment]
Marked as blocked because

> We should probably ask for all of the permissions on first launch

... so I guess it's not really blocked by an external thing, it's a kind of "DON'T MERGE THIS" sticker

**Comment by glacambre** [comment]
This is old and stale. Let's kill it.

**Comment by bovine3dom** [comment]
Reopening as a possible fix for #4247 

We can have a user action from buttons on the help/new tab page

**Pull request closed without merging**